core/src/main/java/net/adoptopenjdk/icedteaweb/resources/downloader/PrioritizedParallelExecutor.java

```diff
@@ -0,0 +1,220 @@
+            }
+        }
+    }
+
+    /**
+     * Wrapper around {@link Future} which allows to construct a chain of futures.
+     * Lower priority futures are canceled if this future is canceled.
+     * If this future fails the result of the lower priority is returned.
+     */
+    private static class ChainableFuture<V> implements Future<V> {
+
+        private final AtomicBoolean terminatedExceptionally = new AtomicBoolean(false);
+        private final CompletableFuture<V> lowerPriorityResult = new CompletableFuture<>();
+        private final AtomicReference<Future<V>> lowerPriority = new AtomicReference<>();
+
+        private final Future<V> delegate;
+
+        private ChainableFuture(Future<V> delegate) {
+            this.delegate = delegate;
+        }
+
+        /**
+         * Set a lower priority to be canceled.
+         * This is not passed in the constructor as the value is not known then.
+         *
+         * @param lowerPriority the lower priority future to be canceled this future is canceled.
+         */
+        void setLowerPriority(Future<V> lowerPriority) {
+            this.lowerPriority.set(lowerPriority);
+
+            if (terminatedExceptionally.get()) {
+                Executors.newSingleThreadExecutor().execute(() -> {
+                            try {
+                                lowerPriorityResult.complete(lowerPriority.get());
+                            } catch (Exception e) {
+                                lowerPriorityResult.completeExceptionally(e);
+                            }
+                        }
+                );
+            }
+        }
+
+        @Override
+        public boolean cancel(boolean mayInterruptIfRunning) {
+            final boolean result = delegate.cancel(mayInterruptIfRunning);
+            cancelLowerPriority();
+            return result;
+        }
+
+        @Override
+        public boolean isCancelled() {
+            return delegate.isCancelled();
+        }
+
+        @Override
+        public boolean isDone() {
+            return delegate.isDone();
+        }
+
+        @Override
+        public V get() throws InterruptedException, ExecutionException {
+            try {
+                return delegate.get();
+            } catch (ExecutionException e) {
+                return returnLowerPriority();
+            }
+        }
+
+        @Override
+        public V get(long timeout, TimeUnit unit) throws InterruptedException, ExecutionException, TimeoutException {
+            final long start = System.nanoTime();
+            try {
+                return delegate.get(timeout, unit);
+            } catch (ExecutionException e) {
+                final long elapsed = System.nanoTime() - start;
+                final long remaining = unit.toNanos(timeout) - elapsed;
+                return returnLowerPriority(remaining, NANOSECONDS);
+            } catch (TimeoutException e) {
+                return returnLowerPriority(1, MILLISECONDS);
```

Comment:
Does this really makes sense to wait longer? Maybe better:

if(lowerPrio.isDone()) {return lowerPrio.get()} else {throw e;}

Next to this calling get() and calling get(long timeout, TimeUnit unit) could now return different results. I think that is a bug and should be changed.

Reply:
Why is this comment resolved? `get()` and `get(long timeout, TimeUnit unit)` could still return different values.

Reply:
I did implement your change request. it checks for lowerPrio done and if so returns that value otherwise throws original exception